pageserver/src/tenant.rs

```diff
@@ -2324,19 +2170,26 @@ impl Tenant {
     /// The returned Timeline is in Loading state. The caller is responsible for
     /// initializing any on-disk state, and for inserting the Timeline to the 'timelines'
     /// map.
+    ///
+    /// `validate_ancestor == false` is used when a timeline is created for deletion
+    /// and we might not have the ancestor present anymore which is fine for to be
+    /// deleted timelines.
     fn create_timeline_struct(
         &self,
         new_timeline_id: TimelineId,
         new_metadata: &TimelineMetadata,
         ancestor: Option<Arc<Timeline>>,
         remote_client: Option<RemoteTimelineClient>,
         init_order: Option<&InitializationOrder>,
+        validate_ancestor: bool,
     ) -> anyhow::Result<Arc<Timeline>> {
-        if let Some(ancestor_timeline_id) = new_metadata.ancestor_timeline() {
-            anyhow::ensure!(
-                ancestor.is_some(),
-                "Timeline's {new_timeline_id} ancestor {ancestor_timeline_id} was not found"
-            )
+        if validate_ancestor {
+            if let Some(ancestor_timeline_id) = new_metadata.ancestor_timeline() {
+                anyhow::ensure!(
+                    new_metadata.ancestor_timeline() == ancestor.as_ref().map(|t| t.timeline_id),
+                    "Timeline's {new_timeline_id} ancestor {ancestor_timeline_id} was not found"
+                );
+            }
```

Comment:
Now the `if let Some` is extra, as the condition uses Option on both sides.

```suggestion
            let tl_id = new_metadata.ancestor_timeline();
            
            anyhow::ensure!(
                tl_id == ancestor.as_ref().map(|t| t.timeline_id),
                "Timeline's {new_timeline_id} ancestor {tl_id:?} was not found"
            );
```

The message is a bit wonky as this checks more than it used to (previously we accepted a `matches!(ancestor, Some(_)) && new_metadata.ancestor_timeline.is_none()`. I don't have a better idea for that, but it's not like we were hitting this on every load.

Reply:
See https://github.com/neondatabase/neon/pull/4552/commits/2386df89cd813f29ef376e12cc971fb7706efebd

> I don't have a better idea for that, but it's not like we were hitting this on every load.

+1

Reply:
In 2386df89cd813f29ef376e12cc971fb7706efebd